src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/TagletWriterImpl.java

```diff
@@ -363,6 +401,199 @@ private boolean isLongOrHasComma(Content c) {
+        assert !ref.isEmpty();
+        DocTree ref0 = ref.get(0);
+        switch (ref0.getKind()) {
+            case TEXT, START_ELEMENT -> {
+                // @see "Reference"
+                // @see <a href="...">...</a>
+                return htmlWriter.commentTagsToContent(element, ref, false, false);
+            }
+
+            case REFERENCE -> {
+                // @see reference label...
+                CommentHelper ch = utils.getCommentHelper(element);
+                String tagName = ch.getTagName(seeTag);
+                String refSignature = ch.getReferencedSignature(ref0);
+                List<? extends DocTree> label = ref.subList(1, ref.size());
+
+                return linkSeeReferenceOutput(element,
+                        seeTag,
+                        refSignature,
+                        ch.getReferencedElement(seeTag),
+                        tagName,
+                        false,
+                        htmlWriter.commentTagsToContent(element, label, context),
+                        (key, args) -> messages.warning(ch.getDocTreePath(seeTag), key, args)
+                );
+            }
+
+            case ERRONEOUS -> {
+                return invalidTagOutput(resources.getText("doclet.tag.invalid_input",
+                                ref0.toString()),
+                        Optional.empty());
+            }
+
+            default -> throw new IllegalStateException(ref0.getKind().toString());
+        }
+
+    }
+
+    /**
+     * Worker method to generate a link from the information in different kinds of tags,
+     * such as {@code {@link ...}} tags, {@code @see ...} tags and the {@code link} markup tag
+     * in a {@code {@snippet ...}} tag.
+     *
+     * @param holder        the element that has the documentation comment containing the information
+     * @param refTree       the tree node containing the information, or {@code null} if not avaiable
+     * @param refSignature  the normalized signature of the target of the reference
+     * @param ref           the target of the reference
+     * @param tagName       the name of the tag in the source, to be used in diagnostics
+     * @param isLinkPlain   {@code true} if the link should be presented in "plain" font,
+     *                      or {@code false} for "code" font.
+     * @param label         the label for the link,
+     *                      or an empty list to use a default label derived from the signature
+     * @param reportWarning a function to report warnings about issues found in the reference
+     *
+     * @return the output containing the generated link, or content indicating an error
+     */
+    Content linkSeeReferenceOutput(Element holder, DocTree refTree, String refSignature, Element ref,
+                                   String tagName, boolean isLinkPlain,
+                                   Content label,
+                                   BiConsumer<String, Object[]> reportWarning) {
```

Comment:
This method is huge: it has 8 parameters and spans 120 lines. We should refactor it in a later PR. For now, consider the below suggestion that changes access level, addresses a few doc typos, and formats parameters one per line.

```suggestion
    /**
     * Worker method to generate a link from the information in different kinds of tags,
     * such as {@code {@link ...}} tags, {@code @see ...} tags and the {@code link} markup tag
     * in a {@code {@snippet ...}} tag.
     *
     * @param holder        the element that has the documentation comment containing the information
     * @param refTree       the tree node containing the information, or {@code null} if not available
     * @param refSignature  the normalized signature of the target of the reference
     * @param ref           the target of the reference
     * @param tagName       the name of the tag in the source, to be used in diagnostics
     * @param isLinkPlain   {@code true} if the link should be presented in "plain" font,
     *                      or {@code false} for "code" font
     * @param label         the label for the link,
     *                      or an empty content to use a default label derived from the signature
     * @param reportWarning a function to report warnings about issues found in the reference
     *
     * @return the output containing the generated link, or content indicating an error
     */
    private Content linkSeeReferenceOutput(Element holder,
                                           DocTree refTree,
                                           String refSignature,
                                           Element ref,
                                           String tagName,
                                           boolean isLinkPlain,
                                           Content label,
                                           BiConsumer<String, Object[]> reportWarning) {
```

Reply:
Yes, it's a big method; my defense is that it is smaller than before!

I'm open to further refactoring if anything obvious becomes apparent.

Reply:
I will make the proposed suggestions.